### PR TITLE
Add address-based index

### PR DESCRIPTION
**Pull request by rnicoll**
Update of #3652 (which is an update of #2802), to latest master. RPC tests are coming, but I wanted to open this now so it's clear there's work going on, make sure we don't duplicate effort.

### CODE REVIEW

File: src/txdb.cpp [hunk 1]
```diff
@@ -163,6 +167,40 @@ bool CBlockTreeDB::WriteTxIndex(const std::vector<std::pair<uint256, CDiskTxPos>
     return WriteBatch(batch);
 }
 
+bool CBlockTreeDB::ReadAddrIndex(const uint160 &addrid, std::vector<CExtDiskTxPos> &list) {
+    CLevelDBIterator *iter = NewIterator();
```
Comment (dexX7) [inline src/txdb.cpp]: It looks like you rebased from the original.

This leaks, you'll need to delete the iterator, or use a smart pointer.

Comment (rnicoll) [inline src/txdb.cpp]: Thanks, added the missing commit in now

**Comment by dcousens** [comment]
concept ACK, will test

**Comment by btcdrak** [comment]
I am very much in favour of a full address indexes as it removes the need for third party applications like insight-api separately parsing the blockchain. However, I'm not sure this PR's approach is the best way. I dont believe it is reorg safe, and there are other issues reported by @gmaxwell https://github.com/CounterpartyXCP/counterparty-lib/issues/783#issuecomment-120130217.

I think overall the approach of #5048 creating a separate index is better (but applied to all addresses rather than just the utxo).

File: src/main.cpp [hunk 2]
```diff
@@ -961,6 +963,45 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
...
+    return true;
+}
+
+bool FindTransactionsByDestination(const CTxDestination &dest, std::set<CExtDiskTxPos> &setpos) {
+    uint160 addrid;
+    const CKeyID *pkeyid = boost::get<CKeyID>(&dest);
+    if (pkeyid) {
+        addrid = static_cast<uint160>(*pkeyid);
+    } else {
+        const CScriptID *pscriptid = boost::get<CScriptID>(&dest);
```
Comment (dcousens) [inline src/main.cpp]: Why not just always use the HASH160 of the script? That way you could query any UTXO.

Comment (afk11) [inline src/main.cpp]: +1 as there are more interesting script types than address types. Address lookup could just produce the necessary output script and look that up. Exposing both would be great.

**Comment by afk11** [comment]
Concept ACK, but needs rebase.

**Comment by gmaxwell** [comment]
Concept (weak)NAK.  (Please don't be upset, I feel that stating my view here is a courtesy over simply saying nothing at all).

I don't see this approach being viable.

There are several interrelated problems, but the foremost of this is that this approach forces its users to use an non-scalable and maintainable method of accessing data in the Bitcoin blockchain-- requiring 50GB+ storage and rapidly growing, where _usually_ a watching wallet is sufficient.  This means that user who do adopt this approach will rapidly be pushed off onto trusting third party API services as the resource costs exceed their tolerance.  

And we're left then with an invasive change that is used by few users and will not be adequately supported (which we've also seen with this change, that no one has even bothered to diagnose the problems with index consistency that this particular implementation has had).

I would much rather see rapid utxo-only rescan for watching import and similar functionality. And for indexing beyond that the indexes should be isolated and modularized to make them maintainable for small user bases, or even for out of tree indexing code.

**Comment by rnicoll** [comment]
Wanted to say, I am following the comments here, I'm just prioritising work elsewhere at the moment. Thanks for all the comments, feedback and analysis everyone, they are all appreciated.

1. Using hash of the full script sounds like an excellent idea.
2. The index consistency issues are likely arising due to lack of recognition of reorgs happening (and therefore the index becomes inconsistent with the blocks on disk).
3. Rapid UTXO-scan sounds like a good first step and we can come back to this.

I _will_ look at this in more depth later, but unlikely before November. If anyone else wants to contribute, PRs against my branch would be much appreciated.

**Comment by dexX7** [comment]
If this patch is going to be overhauled, which I'd like very much, then there are a few points to consider:

1. filtering out of orphaned transactions: similar to the txdb, orphans are kept in the DB, which caused quite a bit confusion in the past. Currently the way to go is to manually check, whether transactions are indeed part of the main chain (e.g. by checking, whether `"confirmations" > 0`), but this may be done on a lower level. Some time ago I extended the RPC to have an option to filter on the RPC level, but this may also not be ideal: https://github.com/dexX7/bitcoin/commit/ffde89049a7ef96b7d09c31cc78a9a366329a24d

2. use more appropriate data types: `verbose` (and `includeorphans`) should be `bool`, not `int`.

3. refine what's actually being indexed: currently a) every data push >=8 bytes,  b) if no such pushes, the entire script is indexed. given that only the lookup by destinations is exposed, it may not be useful to index only the parts that are exposed and needed

4. maybe add additional filters or aggregation calls: for http://redeem.bitwatch.co/ I'm using ["listallunspent"](https://github.com/dexX7/bitcoin/commit/9764daeccfb3f5c813e0a7dd6d022bcd8ffc2edd), which targets unspent outputs (in a somewhat ugly way!), and I've also played with ["getallbalance"](https://github.com/dexX7/bitcoin/commit/813da8e9feb68201ffa2f3b13a27e7802fce9883) to retrieve balance information of arbitrary destinations. It's certainly not very fast, but for me an acceptable middle ground.

5. optionally maintain a mempool based view: DB entries are only written after confirmation, but it would be incredibly useful, if the unconfirmed chain state is available and exposed in one way or another.

Especially to the first and last point a few notes: I'm not sure who else uses the address-index patch, but most notably is probably [counterparty-lib](https://github.com/CounterpartyXCP/counterparty-lib/blob/7ad66f084a02242bbf2cbc1e8bfb9b70387f57b2/counterpartylib/lib/backend/addrindex.py). They are manually filtering the results to get rid of orphans, and they are manually fetching the mempool over and over to get a better understanding of the unconfirmed chain state. This seems incredibly wasteful, performance wise, and doing the work directly in Core may provide a significant performance boost.

**Comment by dcousens** [comment]
After a discussion on IRC with @laanwj.  I've changed my opinion to concept NACK with the conclusion this should be moved to an application external to bitcoind.

**Comment by btcdrak** [comment]
I was also part of the conversations with @laanwj and @dcousens and I'm also in favour of doing this externally to bitcoind.

**Comment by sipa** [comment]
I vote to close this; this should be done in an external index. Perhaps at some point when we can a more modular design it can be supported as an optionally buildable module, but let's not complicate the current database tracking code.

**Comment by arowser** [comment]
Can one of the admins verify this patch?

**Pull request closed without merging**